tests/functional/setups/manu_thermoporomech_nofrac_2d.py

```diff
@@ -0,0 +1,1219 @@
+        """Results object that stores exact and approximated solutions and errors."""
+
+        self.flux_variable: str = "darcy_flux"
+        """Keyword to access the Darcy fluxes."""
+
+        self.stress_variable: str = "poroelastic_force"
+        """Keyword to access the poroelastic force."""
+
+    def set_materials(self):
+        """Set material parameters."""
+        super().set_materials()
+
+        # Instantiate exact solution object after materials have been set
+        self.exact_sol = ManuThermoPoroMechExactSolution2d(self)
+
+    def before_nonlinear_loop(self) -> None:
+        """Update values of external sources."""
+        super().before_nonlinear_loop()
+
+        sd = self.mdg.subdomains()[0]
+        data = self.mdg.subdomain_data(sd)
+        t = self.time_manager.time
+
+        # Mechanics source
+        mech_source = self.exact_sol.mechanics_source(sd=sd, time=t)
+        pp.set_solution_values(
+            name="source_mechanics", values=mech_source, data=data, time_step_index=0
+        )
+
+        # Flow source
+        flow_source = self.exact_sol.flow_source(sd=sd, time=t)
+        pp.set_solution_values(
+            name="source_flow", values=flow_source, data=data, time_step_index=0
+        )
+        # Energy source
+        energy_source = self.exact_sol.energy_source(sd=sd, time=t)
+        pp.set_solution_values(
+            name="source_energy", values=energy_source, data=data, time_step_index=0
+        )
+
+    def _is_nonlinear_problem(self) -> bool:
+        """The problem is non-linear due to the coupling between fluid flux and
+        advective energy transport.
+        """
+        return True
+
+    def bulk_modulus(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
+        """The bulk modulus will be spatially varying if the Lame parameters are so."""
+        x, y = sym.symbols("x y")
+
+        for sd, data in self.mdg.subdomains(dim=self.nd, return_data=True):
+            xc = sd.cell_centers
+            lame_lmbda = sym.lambdify((x, y), self.exact_sol.lame_lmbda, "numpy")(
+                xc[0], xc[1]
+            )
+            lame_mu = sym.lambdify((x, y), self.exact_sol.lame_mu, "numpy")(
+                xc[0], xc[1]
+            )
+
+            bulk = lame_lmbda + (2 / 3) * lame_mu
+            # Special case: If the bulk modulus is constant (no spatial variation), the
+            # lambdify function returns a float or int. In this case, we need to
+            # broadcast the value to all cells.
+            if isinstance(bulk, (float, int)):
+                bulk = bulk * np.ones(sd.num_cells)
+
+            return pp.wrap_as_dense_ad_array(bulk, name="bulk_modulus")
+
+    def set_discretization_parameters(self) -> None:
+        """Set parameters for the subproblems and the combined problem.
+
+        In addition to the standard fields (call to super), the permeability, stiffness
+        parameters, and the Biot and thermal stress tensors are set.
```

Comment:
Wouldn't it be more precise to say that the super call's values are overwritten?